src/JustSaying/Messaging/Middleware/Metrics/StopwatchMiddleware.cs

```diff
@@ -0,0 +1,43 @@
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using JustSaying.Messaging.Middleware.Handle;
+using JustSaying.Messaging.Monitoring;
+
+namespace JustSaying.Messaging.Middleware.Metrics
+{
+    /// <summary>
+    /// This middleware measures the handler's execution duration and reports the results to an <see cref="IMessageMonitor"/>.
+    /// </summary>
+    public class StopwatchMiddleware : MiddlewareBase<HandleMessageContext, bool>
+    {
+        private readonly IMessageMonitor _monitor;
+        private readonly Type _handlerType;
+
+        /// <summary>
+        /// Creates an instance of a <see cref="StopwatchMiddleware"/> that will report results to an
+        /// <see cref="IMessageMonitor"/>.
+        /// </summary>
+        /// <param name="monitor">An <see cref="IMessageMonitor"/> to report results to.</param>
+        /// <param name="handlerType">The type of the handler that results should be reported against.</param>
+        public StopwatchMiddleware(IMessageMonitor monitor, Type handlerType)
+        {
+            _monitor = monitor;
+            _handlerType = handlerType;
+        }
+
+        protected override async Task<bool> RunInnerAsync(HandleMessageContext context, Func<CancellationToken, Task<bool>> func, CancellationToken stoppingToken)
+        {
+            var watch = Stopwatch.StartNew();
+
+            bool result = await func(stoppingToken).ConfigureAwait(false);
+
+            watch.Stop();
```

Comment:
The timing won't be recorded if the func throws, is that intentional?

Reply:
So this is the same logic as the previous handler, [found here](https://github.com/justeat/JustSaying/blob/master/src/JustSaying/Messaging/Monitoring/StopwatchHandler.cs).

I guess it makes sense if you assume that `IMessageMonitor` only calls `HandlerExecutionTime` for successful executions. I'm not sure if that's what I'd expect, but its the way it works now.

Should we change this?

Reply:
If it's consistent with existing behaviour, then that's fine.